### ISSUE TITLE
Error finding a file

### ISSUE THREAD

**Issue by happilymarrieddad** [go-gl/glfw#215]
Hey guys,

This code works just fine in main.go but when I put it in a package it breaks. Does anyone know why? Thanks!

```
go version go1.9.4 linux/amd64
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/nick/Projects/Go"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="/usr/bin/gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build103987606=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOROOT/bin/go version: go version go1.9.4 linux/amd64
GOROOT/bin/go tool compile -V: compile version go1.9.4
uname -sr: Linux 4.10.0-38-generic
Distributor ID:	LinuxMint
Description:	Linux Mint 18.3 Sylvia
Release:	18.3
Codename:	sylvia
/lib/x86_64-linux-gnu/libc.so.6: GNU C Library (Ubuntu GLIBC 2.23-0ubuntu10) stable release version 2.23, by Roland McGrath et al.
gdb --version: GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
```

Makefile
```
CGO=go

run:
	$(CGO) run main.go

build:
	make build_linux
	make build_win
build_win:
	GOOS=windows GOARCH=386 $(CGO) build -o bin/first.exe main.go
build_linux:
	GOOS=linux GOARCH=amd64 $(CGO) build -o bin/first main.go

install:
	make install_deps
	make install_opengl_v4
	make install_glfw
install_deps:
	sudo apt install libgl1-mesa-dev xorg-dev
install_opengl_v4:
	go get -u github.com/go-gl/gl/v4.1-core/gl
install_opengl_v2:
	go get -u github.com/go-gl/gl/v2.1/gl
install_glfw:
	go get -u github.com/go-gl/glfw/v3.2/glfw github.com/go-gl/mathgl/mgl32 golang.org/x/image/colornames github.com/faiface/glhf
```

file
```
package app

import (
	"log"
	"runtime"

	// "github.com/go-gl/gl/v4.1-core/gl"
	"github.com/go-gl/glfw/v3.2/glfw"
)

const (
	title  = "Conway's Game of Life"
	width  = 640
	height = 480
)

type app struct {
	Window *glfw.Window
}

func initGlfw() *glfw.Window {
	log.Println("Initializing GLFW")
	if err := glfw.Init(); err != nil {
		panic(err)
	}

	glfw.WindowHint(glfw.Resizable, glfw.False)
	glfw.WindowHint(glfw.ContextVersionMajor, 4)
	glfw.WindowHint(glfw.ContextVersionMinor, 1)
	glfw.WindowHint(glfw.OpenGLProfile, glfw.OpenGLCoreProfile)
	glfw.WindowHint(glfw.OpenGLForwardCompatible, glfw.True)

	window, err := glfw.CreateWindow(width, height, title, nil, nil)
	if err != nil {
		panic(err)
	}
	window.MakeContextCurrent()

	return window
}

func (a *app) Init() {
	log.Println("Initializing App")
	runtime.LockOSThread()

	a.Window = initGlfw()
}

func (a *app) Start() {
	log.Println("Starting App")
	defer glfw.Terminate()

	for !a.Window.ShouldClose() {

	}
}

func NewApp() app {
	return app{}
}

```

main.go
```
package main

import (
	"./src/system/app"
)

func main() {
	a := app.NewApp()

	a.Init()
	a.Start()
}

```

Folder Structure

![selection_062](https://user-images.githubusercontent.com/6416323/37742034-86b39c34-2d29-11e8-8f0d-baf66bf3b5c5.png)

Error:
![selection_063](https://user-images.githubusercontent.com/6416323/37742041-91e10c4a-2d29-11e8-96f8-38ec15b1aa13.png)

**Comment by happilymarrieddad** [comment]
This works just fine... it seems the same to me
```
package main

import (
	//"log"
	"runtime"

	glfw "github.com/go-gl/glfw/v3.2/glfw"
)

const (
	width  = 500
	height = 500
)

func main() {
	runtime.LockOSThread()

	window := initGlfw()
	defer glfw.Terminate()

	for !window.ShouldClose() {
		// TODO
	}
}

// initGlfw initializes glfw and returns a Window to use.
func initGlfw() *glfw.Window {
	if err := glfw.Init(); err != nil {
		panic(err)
	}

	glfw.WindowHint(glfw.Resizable, glfw.False)
	glfw.WindowHint(glfw.ContextVersionMajor, 4) // OR 2
	glfw.WindowHint(glfw.ContextVersionMinor, 1)
	glfw.WindowHint(glfw.OpenGLProfile, glfw.OpenGLCoreProfile)
	glfw.WindowHint(glfw.OpenGLForwardCompatible, glfw.True)

	window, err := glfw.CreateWindow(width, height, "Conway's Game of Life", nil, nil)
	if err != nil {
		panic(err)
	}
	window.MakeContextCurrent()

	return window
}

```

**Comment by dmitshur** [comment]
This could be one of many things... It's hard to tell right away just from the information you provided so far. Here are a few ideas of what to try to investigate this.

Your `Makefile` does this on first line:

```
CGO=go
```

I wonder if that causes any conflicts...

What happens if you run `go run main.go` directly, skipping the `Makefile`? What if you do `go build`?

There's also a relative import path `"./src/system/app"`. It's legal, but discouraged (see https://golang.org/doc/code.html for advice on how to organize Go code). It might be something to pay attention to when investigating this issue.

You might also want to update to Go 1.10, it's a newer stable release of Go.

**Comment by happilymarrieddad** [comment]
@shurcooL Hey man, sorry for the noob question. I'm pretty new to golang.

How would I import that "package"?
I updated to Golang 1.10 and it fixed it.. that's really weird.. not sure what is different.
I tried go run main.go using Golang 1.9.4 and it didn't make a difference (outside of make)

Hey man, thanks again so much for the reply and the time you took to reply. If you have some time and you don't mind could you please answering the relative path question?

**Comment by happilymarrieddad** [comment]
Do you mean I should import starting with the GOPATH? Like ?
import "graphics/first/src/system/app"

That worked but I'm not sure if that's what you are saying

**Comment by happilymarrieddad** [comment]
@shurcooL Hey man, another super noob question. 

If you are working with a decent sized group of developers, wouldn't relative paths be better because you don't know their environment but you have relative control of the folder structure of the project?? I could be way off base but I was just wondering. Thanks!

**Comment by dmitshur** [comment]
> sorry for the noob question. I'm pretty new to golang.

No problem. I recommend you check out https://golang.org/help/ and join one of the Go communities. For example, you can hop on the Gophers Slack, which has a [#golang-newbies](https://gophers.slack.com/messages/golang-newbies) channel, which is a great place for all beginner questions.

> How would I import that "package"? I updated to Golang 1.10 and it fixed it.. that's really weird.. not sure what is different. I tried go run main.go using Golang 1.9.4 and it didn't make a difference (outside of make)

I see. It sounds like it might've been a bug in Go 1.9, and it's fixed in Go 1.10. You had a slightly unusual setup, so that's likely why there was a problem with it in 1.9.

> Do you mean I should import starting with the GOPATH? Like ? import "graphics/first/src/system/app"

A relative import path is one that begins with `.` or `..`. E.g., "./foo/bar". It imports a package in a directory relative to the current package's directory. This is very infrequently done in normal Go code.

A (normal) import path is one that doesn't begin with `.` nor `..`. E.g., "net/http" or "github.com/user/repo" are normal import paths.

From your first post where you posted `go env` output, I can see your GOPATH is `/home/nick/Projects/Go`. Packages are placed in `GOPATH/src/import/path`, so if your `app` directory is at `/home/nick/Projects/Go/src/graphics/first/src/system/app`, then "graphics/first/src/system/app" is indeed the right import path.

In the end, I highly recommend you read https://golang.org/doc/code.html and/or watch the https://www.youtube.com/watch?v=XCsL89YtqCs screencast (5 min). It explains the proper way to organize Go code inside a GOPATH workspace.

> If you are working with a decent sized group of developers, wouldn't relative paths be better because you don't know their environment but you have relative control of the folder structure of the project?? I could be way off base but I was just wondering.

This is a common first impression that people get when they're just learning about Go. In practice, normal import paths are used most of the time, and they work well even when multiple people are working on a project. They're always resolved against GOPATH/src, so it's very predictable and easy to find each Go package. There are particular workflows that you will understand with a bit of practice that make normal import paths very pleasant to work with. You can check out https://github.com/google/go-github or https://github.com/upspin/upspin to see some high quality Go projects with many collaborators and no relative import paths.

Hope that helps, and good luck!

**Comment by happilymarrieddad** [comment]
@shurcooL thanks so much! I can't wait to get my projects going!!!
